### PR TITLE
DEPR: Enforce numeric_only=False in groupby sum/mean

### PR DESCRIPTION
**Pull request by rhshadrach**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Targeting these two ops first because they're used as an example op in a lot of tests. Hoping the rest of the ops are easier. Note this doesn't check off the corresponding line in the list of deprecations.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
is the failed asv time_pivot_table_margins_only_column a false-positive?

File: pandas/tests/groupby/test_groupby.py [hunk 1]
```diff
@@ -1170,19 +1175,17 @@ def test_groupby_with_hier_columns():
...
 
-    msg = "The default value of numeric_only"
-    with tm.assert_produces_warning(FutureWarning, match=msg):
-        result = grouped.sum()
-        expected = df.groupby("A").sum()
+    result = grouped.sum()
+    expected = df.groupby(df["A"].rename("key")).sum()
     tm.assert_frame_equal(
         result, expected, check_names=False
     )  # Note: no names when grouping by value
```
Comment (MarcoGorelli) [inline pandas/tests/groupby/test_groupby.py:1191]: would
```suggestion
    expected = df.groupby(df["A"].rename(None)).sum()
    tm.assert_frame_equal(result, expected)
```
work?

File: pandas/tests/groupby/test_pipe.py [hunk 2]
```diff
@@ -55,25 +55,22 @@ def test_pipe_args():
     )
 
     def f(dfgb, arg1):
-        return dfgb.filter(lambda grp: grp.y.mean() > arg1, dropna=False).groupby(
-            dfgb.grouper
-        )
+        filtered = dfgb.filter(lambda grp: grp.y.mean() > arg1, dropna=False)
+        return filtered.groupby("group")
```
Comment (MarcoGorelli) [inline pandas/tests/groupby/test_pipe.py:59]: just for my understanding, why did this need to change?

**Review by MarcoGorelli — COMMENTED** [review]
Generally looks good, just got minor comments / a question

File: pandas/tests/groupby/test_pipe.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/tests/groupby/test_pipe.py:59]: When using `.groupby(dfgb.grouper)`, the column `group` in included in the operation. While the `sum` is successful, the function `g` then adds `arg2` (with value 10) and this fails. We could also supply `numeric_only=True` to the two `sum` operations in `g`, but this seems a little easier.

**Comment by rhshadrach** [comment]
Thanks @jbrockmendel - ASV failure was due to trying to take the mean of a string column; I added it to the `columns` instead. It now takes 30ms on my machine.

File: pandas/tests/groupby/test_groupby.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/tests/groupby/test_groupby.py]: Yep - thanks!

File: pandas/tests/extension/base/groupby.py [hunk 3]
```diff
@@ -100,17 +106,23 @@ def test_in_numeric_groupby(self, data_for_grouping):
...
             expected = pd.Index(["B", "C"])
         else:
             expected = pd.Index(["C"])
 
+        if "B" in expected:
+            result = df.groupby("A").sum().columns
+        else:
+            with pytest.raises(TypeError, match="does not support"):
+                df.groupby("A").sum().columns
+            result = df.groupby("A").sum(numeric_only=True).columns
```
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/tests/extension/base/groupby.py]: Yep - thanks!

**Review by mroeschke — APPROVED** [review]
LGTM cc @MarcoGorelli merge when ready

**Review by MarcoGorelli — APPROVED** [review]
looks good to me, thanks @rhshadrach !

**Pull request merged**